### PR TITLE
[Merged by Bors] - Add `SensitiveUrl` to redact user secrets from endpoints

### PR DESCRIPTION
**Pull request by macladson**
## Issue Addressed

#2276 

## Proposed Changes

Add the `SensitiveUrl` struct which wraps `Url` and implements custom `Display` and `Debug` traits to redact user secrets from being logged in eth1 endpoints, beacon node endpoints and metrics.

## Additional Info

This also includes a small rewrite of the eth1 crate to make requests using `Url` instead of `&str`. 
Some error messages have also been changed to remove `Url` data.

### CODE REVIEW

File: beacon_node/eth1/src/service.rs [hunk 1]
```diff
@@ -140,7 +149,7 @@ async fn endpoint_state(
         warn!(
             log,
             "Error connecting to eth1 node endpoint";
-            "endpoint" => endpoint,
+            "endpoint" => endpoint.to_string(),
```
Comment (michaelsproul) [inline beacon_node/eth1/src/service.rs]: We should use the magic `%endpoint` syntax for the `Display` implementation (there are a few in this file)

```suggestion
            "endpoint" => %endpoint,
```

File: common/sensitive_url/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+use url::Url;
+
+#[derive(Debug)]
+pub enum SensitiveError {
+    InvalidUrl(String),
+    ParseError(url::ParseError),
+    RedactError(String),
+}
+
+// Wrapper around Url which provides a custom `Display` implemenation to protect user secrets.
```
Comment (michaelsproul) [inline common/sensitive_url/src/lib.rs]:
```suggestion
// Wrapper around Url which provides a custom `Display` implementation to protect user secrets.
```

File: beacon_node/eth1/tests/test.rs [hunk 3]
```diff
@@ -53,7 +54,7 @@ fn random_deposit_data() -> DepositData {
 /// Blocking operation to get the deposit logs from the `deposit_contract`.
 async fn blocking_deposit_logs(eth1: &GanacheEth1Instance, range: Range<u64>) -> Vec<Log> {
     get_deposit_logs_in_range(
-        &eth1.endpoint(),
+        &SensitiveUrl::parse(eth1.endpoint().as_str()).unwrap(),
```
Comment (michaelsproul) [inline beacon_node/eth1/tests/test.rs:57]: This is just a matter of style (and I don't suggest changing it), but often you can convert `String` -> `&str` by writing `&`, like `&eth1.endpoint()`.

**Review by michaelsproul — COMMENTED** [review]
This is great, really well implemented.

Just a few minor nitpicks to address and then I think we're ready to merge :tada:

File: beacon_node/eth1/src/service.rs (hunk 1 above)
Comment (macladson) [reply beacon_node/eth1/src/service.rs]: Ahh yes, I had forgotten about %. Thanks!

**Review by michaelsproul — APPROVED** [review]
Ready to merge with or after #2331

**Comment by michaelsproul** [comment]
ah, merge conflict. I'll cancel this and let the audit fix go through first

bors r-

**Comment by michaelsproul** [comment]
let's try it

bors r+

**Pull request closed without merging**